File: src/org/candango/carcara/engine/AbstractDatabaseLoader.java
package org.candango.carcara.engine;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.ArrayList;

import org.candango.carcara.engine.mysql.MysqlDatabaseLoader;
import org.candango.carcara.engine.pgsql.PgsqlDatabaseLoader;
import org.candango.carcara.model.database.Table;

public abstract class AbstractDatabaseLoader implements DatabaseLoader {
	
	public static final int MYSQL_DATABASE = 1;
	
	public static final int PGSQL_DATABASE = 2;
	
	private Connection conn;
	
	private ArrayList<Table> tableList = new ArrayList<Table>();
	
	public static DatabaseLoader getLoader( int witchLoader ) {
		
		switch( witchLoader ) {
			case( MYSQL_DATABASE ):
				return new MysqlDatabaseLoader();
			case( PGSQL_DATABASE ):
				return new PgsqlDatabaseLoader();
		}
		
		return null;
	}
	
	public Connection getConnection(){
		return conn;
	}
	
	public void setConnection( Connection conn ) {
		this.conn = conn;
	}
	
	public void connect( DatabaseConfiguration configuration ) {
		System.out.println( "Connecting..." );
		try {
			Class.forName( configuration.getDriver() );
			
			setConnection( DriverManager.getConnection( 
					configuration.getUrl(),
					configuration.getUser(), configuration.getPassword() ) );
			
		} catch (ClassNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (SQLException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		System.out.println( "Oukie!!" );
		
	}
	
	public void disconnect() {
		try {
			getConnection().close();
		} catch (SQLException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	public void doLoad( DatabaseConfiguration configuration ) {
		
		loadTables( configuration );
		
		for( Table table : getTables() ) {
			loadFields( configuration, table );
		}
		
	}
	
	protected abstract void loadTables( DatabaseConfiguration configuration );
	
	protected abstract void loadFields( DatabaseConfiguration configuration, Table table );

	protected void addTable( Table table ) {
		tableList.add( table );
	}
	
	public Table[] getTables() {
		Table[] tables = new Table[ tableList.size() ];
		return tableList.toArray( tables );
	}
	
}

File: src/Carcara/Engine/AbstractDatabaseLoader.php
<?php
/**
 * Carcara (http://carcara.candango.org)
 *
 * @link      http://github.com/candango/carcara
 * @copyright Copyright (c) 2018 Flavio Garcia
 * @license   https://www.apache.org/licenses/LICENSE-2.0  Apache-2.0
 */

namespace Candango\Carcara\Engine
{

    use Candango\Carcara\Conf;
    use Candango\Carcara\Engine\Mysql\MysqlDatabaseLoader;
    use Candango\Carcara\Engine\Pgsql\PgsqlDatabaseLoader;
    use Candango\Carcara\Model\Database\Table;

    abstract class AbstractDatabaseLoader implements DatabaseLoader
    {
        const MYSQL = "mysql";

        const PGSQL = "pgsql";

        /**
         * Dao Factory Connection
         *
         * @var \PDO
         */
        private $conn;


        /**
         * @var Conf
         */
        private $conf;

        private $tables = array();

        public function __construct($conf)
        {
            $this->setConf($conf);
        }

        /**
         * @param Conf $conf
         * @return MysqlDatabaseLoader|null
         */
        public static function getLoader(Conf $conf)
        {
            switch ($conf->getType()) {
                case(self::MYSQL):
                    return new MysqlDatabaseLoader($conf);
                    break;
                case(self::PGSQL):
                    return new PgsqlDatabaseLoader($conf);
                    break;
            }

            return null;
        }

        public function getConnection()
        {
            return $this->conn;
        }

        public function setConnection($conn)
        {
            $this->conn = $conn;
        }

        /**
         * Set the pdo connection to null in order to get the connection closed
         */
        public function disconnect()
        {
            $this->conn = null;
        }

        /**
         * @return Conf
         */
        public function getConf()
        {
            return $this->conf;
        }

        /**
         * @param Conf $conf
         */
        public function setConf(Conf $conf)
        {
            $this->conf = $conf;
        }

        /**
         *
         * @return void
         */
        public function doLoad()
        {
            $this->loadTables();

            foreach ($this->getTables() as $table){
                $this->loadFields($table);
            }
        }

        protected abstract function loadTables();

        protected abstract function loadFields(Table $table);

        /**
         *
         * @param Table $table
         */
        protected function addTable(Table $table)
        {
            $this->tables[] = $table;
        }

        /**
         * @return array of Table
         */
        public function getTables()
        {
            return $this->tables;
        }
    }
}
